source/extensions/clusters/dynamic_forward_proxy/cluster.cc

```diff
@@ -267,44 +264,31 @@ void Cluster::addOrUpdateHost(
       //                     semantics, meaning the cache would expose multiple addresses and the
       //                     cluster would create multiple logical hosts based on those addresses.
       //                     We will leave this is a follow up depending on need.
-      ASSERT(host_info == host_map_it->second.shared_host_info_);
       ASSERT(host_map_it->second.shared_host_info_->address() !=
              host_map_it->second.logical_host_->address());
-      ENVOY_LOG(debug, "updating dfproxy cluster host address '{}'", host);
-      host_map_it->second.logical_host_->setNewAddresses(
-          host_info->address(), host_info->addressList(), dummy_lb_endpoint_);
-      return;
-    }
 
-    ENVOY_LOG(debug, "adding new dfproxy cluster host '{}'", host);
+      // remove the old host
+      hosts_removed.emplace_back(host_map_it->second.logical_host_);
+      ENVOY_LOG(debug, "updating dfproxy cluster host address '{}'", host);
+      host_map_.erase(host_map_it);
+      host_map_.try_emplace(host, host_info, new_host);
 
-    emplaced_host = host_map_
-                        .try_emplace(host, host_info,
-                                     std::make_shared<Upstream::LogicalHost>(
-                                         info(), std::string{host}, host_info->address(),
-                                         host_info->addressList(), dummy_locality_lb_endpoint_,
-                                         dummy_lb_endpoint_, nullptr, time_source_))
-                        .first->second.logical_host_;
+    } else {
+      ENVOY_LOG(debug, "adding new dfproxy cluster host '{}'", host);
+      host_map_.try_emplace(host, host_info, new_host);
+    }
+    hosts_added.emplace_back(new_host);
   }
 
-  ASSERT(emplaced_host);
-  if (hosts_added == nullptr) {
-    hosts_added = std::make_unique<Upstream::HostVector>();
-  }
-  hosts_added->emplace_back(emplaced_host);
+  ASSERT(hosts_added.size() > 0);
+  updatePriorityState(hosts_added, hosts_removed);
```

Comment:
Actually looking through this code belatedly, doesn't updatePriorityState cause updateHosts cause drainOrCloseConnPools(host, ConnectionPool::DrainBehavior::DrainAndDelete) on the "old host"?  I think this could be a pretty bad latency hit where we want to lame-duck drain rather than force-remove drain.

Reply:
Oh, I see the problem.
But, I think it might be reasonable to force-remove drain.
When dns resolve result changes, force-remove drain the `old host` should be expected?

The root cause is the dfp cluster use logical DNS mode instead of strict dns mode, it may lead to host change even when dns resolve result not changes.

I haven't found a simple solution yet.
Maybe we can recommend people to use the new [sub_cluster_config](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/dynamic_forward_proxy/v3/dynamic_forward_proxy.proto#envoy-v3-api-field-extensions-filters-http-dynamic-forward-proxy-v3-filterconfig-sub-cluster-config)?
Or, any suggestions?

Reply:
DrainExistingConnections would maybe be OK, but DrainAndDelete is too aggressive IMO as it will force-close connections.

Either way I'm thinking we should probably roll this back as I don't think the prior leak had caused many problems in production and (I think) this ended up being an unexpected non-flag-guarded behavioral change.

I am also OK writing a test to verify if this resulted in the behavioral change I think it did before we go that route but  @yanavlasov @ggreenway for their thoughts.

Reply:
This bug is first reported in #30999, and it leads to a large leak, according to the profile data.
But, according @rbtz-openai's update, brotli decompression is the root cause for the large leak.
I also agree that it just leaking in extremely case, with very low rate.
Revert this patch might be a proper choice, since it's hard to keep the old behaviour to fix this extremely low rate leak.
Maybe @rbtz-openai could reconfirm that this dfp leak could be ignored in your production case?

For "DrainAndDelete will force-close connections":
After reading the comments of DrainExistingConnections/DrainAndDelete, and the code of drainConnections(drain_behavior), I didn't find force-close connections.
I think I missed it somewhere. Would you please give a guideline? i.e. some keywords of the code.
Thanks very much~

Reply:
Yeah so I added tests (https://github.com/envoyproxy/envoy/pull/32710) and while I'm not sure why (having looked at the code) we're not force-closing, it does appear we're not force closing. I'll do a little more investigation before pushing for rollback.  Thanks for looking into this as well!

Reply:
Ah ok, figured out what's wrong with my test.

OK so this PR both did fix the memory leak described, but also result in a functional change to connection lifetimes.

By the Envoy change guidelines I think it should not have landed without flag protection (because data plane change).  The particular change is problematic for Envoy Mobile.  Given the memory leak is long standing and hadn't been largely I'm inclined to roll back to prior (buggy) state than keep this (also buggy for E-M) state.  We could also flag guard the change if there's anyone who wants the memory leak fix and doesn't mind the connectivity change?

cc @yanavlasov @RyanTheOptimist for second opinions.

for context, this fixes the leak in the mutable_cross_priority_host_map_  as described but does so by updating the host in worker threads, which results in an (new, undesirable for E-M) connection pool drain and fresh non-warmed-CWND connection being used for subsequent streams.

Ideally we'd have a solution which handled both but until we have a solution the question is which bug to prefer. I'm inclined to stick with the old one but would also be amenable into making it a flag-guarded "behavioral change + fix".  Unclear if the correct behavior for a server is to drain on DNS refresh - it does do a better job of respecting DNS TTL but can also result in a lot of connection churn and suboptimal bandwidth.

Reply:
Yeah, I agree with @alyssawilk that since this included a non-flag protected behavior change (especially since it broke mobile), we should roll this back. We can go ahead and roll forward with flag protection, of course, but we should return to a stable state first, I think.

Reply:
okay, I see. Rolling back does make sense. If someone really need this fix, we could create a new PR with a flag guard.